### PR TITLE
Get IO.FileSystem tests running on NetFX

### PR DESCRIPTION
**Pull request by ianhays**
There were a high number of failing FileSystem tests for NetFX. This PR modifies the tests to run on NetFX without failures and adds comments for the reasons for the tests that had to be disabled.

resolves https://github.com/dotnet/corefx/issues/14987
resolves https://github.com/dotnet/corefx/issues/17659

cc: @stephentoub @JeremyKuhne @safern

### CODE REVIEW

File: src/System.IO.FileSystem/tests/Directory/GetFileSystemEntries_str_str.cs [hunk 1]
```diff
@@ -298,6 +294,36 @@ public void WindowsSearchPatternInvalid()
         }
 
         [Fact]
+        [PlatformSpecific(TestPlatforms.Windows)]  // Windows-invalid search patterns throw
+        [SkipOnTargetFramework(TargetFrameworkMonikers.NetFramework, "")]
```
Comment (safern) [inline src/System.IO.FileSystem/tests/Directory/GetFileSystemEntries_str_str.cs]: Missing comment to explain why this is disabled.

File: src/System.IO.FileSystem/tests/FileInfo/Open.cs [hunk 2]
```diff
@@ -12,6 +12,40 @@ protected override FileStream CreateFileStream(string path, FileMode mode)
         {
             return new FileInfo(path).Open(mode);
         }
+
+        [Fact]
+        [SkipOnTargetFramework(TargetFrameworkMonikers.NetFramework, "FileInfo.Open(string, filemode) on netfx always uses FileAccess.ReadWrite instead of choosing a FileAccess based on the FileMode. This bug was fixed in netcoreapp.")]
```
Comment (safern) [inline src/System.IO.FileSystem/tests/FileInfo/Open.cs]: Should we open an issue to track this difference and mark it as netfx-port-consider?

File: src/System.IO.FileSystem/tests/FileStream/WriteAsync.cs [hunk 3]
```diff
@@ -207,6 +207,7 @@ public void CancelledTokenFastPath()
         }
 
         [Fact]
+        [SkipOnTargetFramework(TargetFrameworkMonikers.NetFramework, "In netcoreapp we modified ReadAsync/WriteAsync to complete synchronously here, but that change was not backported to netfx.")]
```
Comment (safern) [inline src/System.IO.FileSystem/tests/FileStream/WriteAsync.cs]: Same here; netfx-port-consider issue?

File: src/System.IO.FileSystem/tests/FileStream/ReadAsync.cs [hunk 4]
```diff
@@ -226,6 +226,7 @@ public void CancelledTokenFastPath()
         }
 
         [Fact]
+        [SkipOnTargetFramework(TargetFrameworkMonikers.NetFramework, "In netcoreapp we modified ReadAsync/WriteAsync to complete synchronously here, but that change was not backported to netfx.")]
```
Comment (safern) [inline src/System.IO.FileSystem/tests/FileStream/ReadAsync.cs]: Same here as comment above.

File: src/System.IO.FileSystem/tests/FileStream/IsAsync.cs [hunk 5]
```diff
@@ -40,6 +40,7 @@ public void FileOptionsAsynchronousConstructorArg()
         }
 
         [Fact]
+        [SkipOnTargetFramework(TargetFrameworkMonikers.NetFramework, "NetFX defaults useAsync to false for the SafeFileHandle/FileAccess constructor which leads to an ArgumentException for an async handle. In netcoreapp this constructor was fixed to check the async status of the handle and use that for the useAsync value.")]
```
Comment (safern) [inline src/System.IO.FileSystem/tests/FileStream/IsAsync.cs]: Same here; should we open an issue as netfx-port-consider or if there is one already open should we include it in the comment?

File: src/System.IO.FileSystem/tests/FileStream/CopyToAsync.cs [hunk 6]
```diff
@@ -47,6 +42,53 @@ public void InvalidArgs_Throws(bool useAsync)
         [Theory]
         [InlineData(false)]
         [InlineData(true)]
+        [SkipOnTargetFramework(TargetFrameworkMonikers.NetFramework, "This fails on netcoreapp because the Stream CopyToAsync calls Length which checks the validity of the underlying handle. On NetFX the operation no-ops for no input or delays failure to execution for input. See /dotnet/coreclr/pull/4540.")]
```
Comment (safern) [inline src/System.IO.FileSystem/tests/FileStream/CopyToAsync.cs]: Nit: the comment is a bit confusing as it starts saying "this fails on netcoreapp" and it is being disabled for netfx.

File: src/System.IO.FileSystem/tests/Directory/GetFileSystemEntries_str_str.cs (hunk 1 above)
Comment (ianhays) [inline src/System.IO.FileSystem/tests/Directory/GetFileSystemEntries_str_str.cs]: Oops, thanks.

File: src/System.IO.FileSystem/tests/FileInfo/Open.cs (hunk 2 above)
Comment (ianhays) [inline src/System.IO.FileSystem/tests/FileInfo/Open.cs]: I don't personally think we should. There are too many inconsistencies requiring too many bugs to open across the framework. And netfx-port-consider is generally applied to closed issues which we do not have for these  bugs, so I guess we would open new issues just to close them and mark them? Seems like too much. I think just having this tag and the associated comment is adequate for now, and if/when we make an IO backport push we can easily find the bugs that have been fixed in netcoreapp but not netfx by searching for these attributes.

File: src/System.IO.FileSystem/tests/FileStream/CopyToAsync.cs (hunk 6 above)
Comment (ianhays) [inline src/System.IO.FileSystem/tests/FileStream/CopyToAsync.cs]: Good point. I was referring to the CopyToAsync call by "This fails" which is confusing. Fixed.

File: src/System.IO.FileSystem/tests/FileInfo/Open.cs (hunk 2 above)
Comment (danmosemsft) [inline src/System.IO.FileSystem/tests/FileInfo/Open.cs]: I agree, it's not worth creating issues just to track ports. We've made very many fixes to Core already  that we haven't tracked so when we get to that point we couldn't rely on issues anyway. Exception if it's something critical enough we might service desktop for it. Then I'd email Alex G.

**Comment by danmosemsft** [comment]
@ianhays feel free to merge, but are you still comfortable with the breaking changes that made these changes necessary? These decisions were presumably made before we became more serious about desktop compatibility. We certainly don't have a goal of mirroring every idiosyncracy of desktop though, it's a balance.

**Comment by safern** [comment]
@danmosemsft should we merge? @ianhays is OOF for 2 weeks and he didn't responded on the last question.

**Comment by danmosemsft** [comment]
@JeremyKuhne or @stephentoub can one of you OK these changes from NETFX also? thought I guess Jan did.

**Comment by ianhays** [comment]
> are you still comfortable with the breaking changes that made these changes necessary? These decisions were presumably made before we became more serious about desktop compatibility

I'd say so, yes. Most of these differences are due to bugs in netfx that were fixed in netcoreapp. I'll let @JeremyKuhne speak to the change to wildcard characters, but otherwise I think we're in a good state here.  I'm going to merge and we can always modify the test if we decide to modify the behavior of the wildcards later, though that change was recent so I don't see us wanting to revert it.

**Pull request merged**